File: everything_parsing/30-shift-reduce-handles/java/ShiftReduceTest.java
import java.util.Arrays;
import java.util.List;

/** java -cp java/out ShiftReduceTest */
public final class ShiftReduceTest {

    static Grammar expr() {
        return Grammar.parse("""
            %start E
            E -> E + T | T
            T -> T * F | F
            F -> ( E ) | id
            """);
    }
    static List<String> toks(String s) { return Arrays.asList(s.split(" ")); }

    public static void main(String[] args) {
        System.out.println("ShiftReduceTest");

        Grammar g = expr();
        ShiftReduce sr = new ShiftReduce(g);

        // ── a full parse ──
        ShiftReduce.Result r = sr.parse(toks("id + id * id"));
        Assert.that(r.ok(), "id + id * id parses bottom-up");
        Assert.that(!r.conflict(), "unambiguous grammar: no shift-reduce conflict");
        Assert.equals(r.steps().get(r.steps().size() - 1).action(), "ACCEPT", "ends in ACCEPT");

        // ── reductions reversed = a rightmost derivation of the input ──
        List<String> form = sr.rightmostDerivation(r.reductions());
        Assert.equals(form, toks("id + id * id"),
            "reversing the reductions yields a rightmost derivation ending at the input");

        // ── the LAST reduction is the FIRST step of the rightmost derivation: E -> E + T ──
        Assert.equals(ShiftReduce.compact(r.reductions().get(r.reductions().size() - 1)), "E -> E + T",
            "the final reduction corresponds to the top of the rightmost derivation");

        // ── * binds tighter: T * F is reduced before E + T ──
        List<String> red = new java.util.ArrayList<>();
        for (Grammar.Production p : r.reductions()) red.add(ShiftReduce.compact(p));
        Assert.that(red.indexOf("T -> T * F") < red.indexOf("E -> E + T"),
            "T -> T * F is reduced before E -> E + T (precedence shows up in reduction order)");

        // ── parens ──
        ShiftReduce.Result r2 = sr.parse(toks("( id + id ) * id"));
        Assert.that(r2.ok(), "( id + id ) * id parses");
        Assert.equals(sr.rightmostDerivation(r2.reductions()), toks("( id + id ) * id"), "and reverses correctly");
        boolean sawParenHandle = r2.steps().stream().anyMatch(s -> s.action().contains("F -> ( E )"));
        Assert.that(sawParenHandle, "( E ) is reduced as a handle");

        // ── single id ──
        ShiftReduce.Result r3 = sr.parse(toks("id"));
        Assert.equals(r3.reductions().size(), 3, "id needs 3 reductions: F->id, T->F, E->T");

        // ── ambiguous grammar: shift-reduce conflict ──
        Grammar amb = Grammar.parse("""
            %start E
            E -> E + E | id
            """);
        ShiftReduce sr2 = new ShiftReduce(amb);
        ShiftReduce.Result ra = sr2.parse(toks("id + id + id"));
        Assert.that(ra.ok(), "the ambiguous grammar still finds A parse");
        Assert.that(ra.conflict(), "and reports a shift-reduce conflict");
        Assert.that(ra.conflictNote().contains("E + E"), "the conflict state stack is E + E");
        Assert.that(ra.conflictNote().contains("shift") && ra.conflictNote().contains("reduce"),
            "both shift and reduce lead to accept");

        Assert.summary();
    }
}

File: everything_parsing/30-shift-reduce-handles/js/shiftreduce.test.mjs
// node js/shiftreduce.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { ShiftReduce, compact } from "./shiftreduce.mjs";

console.log("shiftreduce.test.mjs");

const expr = () => Grammar.parse(`
%start E
E -> E + T | T
T -> T * F | F
F -> ( E ) | id
`);
const toks = s => s.split(" ");
const eqArr = (a, b) => a.length === b.length && a.every((x, i) => x === b[i]);

// helper mirroring java rightmostDerivation
function rightmost(g, reductions) {
  let form = [g.start];
  for (const p of [...reductions].reverse()) {
    let i = -1;
    for (let k = form.length - 1; k >= 0; k--) if (form[k] === p.lhs) { i = k; break; }
    form = [...form.slice(0, i), ...p.rhs, ...form.slice(i + 1)];
  }
  return form;
}

const g = expr();
const sr = new ShiftReduce(g);

const r = sr.parse(toks("id + id * id"));
that(r.ok, "id + id * id parses bottom-up");
that(!r.conflict, "unambiguous grammar: no shift-reduce conflict");
equals(r.steps[r.steps.length - 1].action, "ACCEPT", "ends in ACCEPT");

that(eqArr(rightmost(g, r.reductions), toks("id + id * id")),
  "reversing the reductions yields a rightmost derivation ending at the input");
equals(compact(r.reductions[r.reductions.length - 1]), "E -> E + T",
  "the final reduction corresponds to the top of the rightmost derivation");

const red = r.reductions.map(compact);
that(red.indexOf("T -> T * F") < red.indexOf("E -> E + T"),
  "T -> T * F is reduced before E -> E + T");

const r2 = sr.parse(toks("( id + id ) * id"));
that(r2.ok, "( id + id ) * id parses");
that(eqArr(rightmost(g, r2.reductions), toks("( id + id ) * id")), "and reverses correctly");
that(r2.steps.some(s => s.action.includes("F -> ( E )")), "( E ) is reduced as a handle");

const r3 = sr.parse(toks("id"));
equals(r3.reductions.length, 3, "id needs 3 reductions");

const amb = Grammar.parse(`
%start E
E -> E + E | id
`);
const ra = new ShiftReduce(amb).parse(toks("id + id + id"));
that(ra.ok, "the ambiguous grammar still finds A parse");
that(ra.conflict, "and reports a shift-reduce conflict");
that(ra.conflictNote.includes("E + E"), "the conflict state stack is E + E");
that(ra.conflictNote.includes("shift") && ra.conflictNote.includes("reduce"),
  "both shift and reduce lead to accept");

summary();
